File: disaster_relief_app/lib/models/drip_models.dart
// ignore_for_file: invalid_annotation_target

import 'package:freezed_annotation/freezed_annotation.dart';

part 'drip_models.freezed.dart';
part 'drip_models.g.dart';

@freezed
class LocalizedText with _$LocalizedText {
  const factory LocalizedText({
    @JsonKey(name: 'zh-TW') required String zhTW,
    @JsonKey(name: 'en-US') required String enUS,
  }) = _LocalizedText;

  factory LocalizedText.fromJson(Map<String, dynamic> json) => _$LocalizedTextFromJson(json);
}

@freezed
class GeoLocation with _$GeoLocation {
  const factory GeoLocation({
    required double lat,
    required double lng,
    required String address,
    String? geohash,
  }) = _GeoLocation;

  factory GeoLocation.fromJson(Map<String, dynamic> json) => _$GeoLocationFromJson(json);
}

@freezed
class MissionRequirements with _$MissionRequirements {
  const factory MissionRequirements({
    @Default(<String>[]) List<String> materials,
    @JsonKey(name: 'manpower_needed') @Default(0) int manpowerNeeded,
    @JsonKey(name: 'manpower_current') @Default(0) int manpowerCurrent,
  }) = _MissionRequirements;

  factory MissionRequirements.fromJson(Map<String, dynamic> json) => _$MissionRequirementsFromJson(json);
}

enum MissionStatus {
  @JsonValue('open')
  open,
  @JsonValue('in_progress')
  inProgress,
  @JsonValue('done')
  done,
  @JsonValue('canceled')
  canceled,
}

enum MissionType { rescue, medical, transport, clearing, other }

@freezed
class Mission with _$Mission {
  const factory Mission({
    required String id,
    required LocalizedText title,
    LocalizedText? description,
    @Default(<MissionType>[]) List<MissionType> types,
    @Default(MissionStatus.open) MissionStatus status,
    @Default(false) bool priority,
    @Default(MissionRequirements()) MissionRequirements requirements,
    required GeoLocation location,
    @JsonKey(name: 'contact_name') String? contactName,
    @JsonKey(name: 'contact_phone_masked') String? contactPhoneMasked,
    @JsonKey(name: 'creator_id') required String creatorId,
    @JsonKey(name: 'chat_room_id') String? chatRoomId,
    @JsonKey(name: 'participant_count') @Default(0) int participantCount,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _Mission;

  factory Mission.fromJson(Map<String, dynamic> json) => _$MissionFromJson(json);
}

@freezed
class MissionParticipant with _$MissionParticipant {
  const factory MissionParticipant({
    @JsonKey(name: 'mission_id') required String missionId,
    @JsonKey(name: 'user_id') required String userId,
    @JsonKey(name: 'is_visible') @Default(true) bool isVisible,
    @JsonKey(name: 'joined_at') DateTime? joinedAt,
  }) = _MissionParticipant;

  factory MissionParticipant.fromJson(Map<String, dynamic> json) => _$MissionParticipantFromJson(json);
}

@freezed
class MissionMessage with _$MissionMessage {
  const factory MissionMessage({
    required String id,
    @JsonKey(name: 'mission_id') required String missionId,
    @JsonKey(name: 'author_id') required String authorId,
    required String content,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'expires_at') DateTime? expiresAt,
  }) = _MissionMessage;

  factory MissionMessage.fromJson(Map<String, dynamic> json) => _$MissionMessageFromJson(json);
}

@freezed
class ShuttleVehicleInfo with _$ShuttleVehicleInfo {
  const factory ShuttleVehicleInfo({
    required String type,
    String? plate,
  }) = _ShuttleVehicleInfo;

  factory ShuttleVehicleInfo.fromJson(Map<String, dynamic> json) => _$ShuttleVehicleInfoFromJson(json);
}

@freezed
class ShuttleRoute with _$ShuttleRoute {
  const factory ShuttleRoute({
    required GeoLocation origin,
    required GeoLocation destination,
    @Default(<GeoLocation>[]) List<GeoLocation> stops,
  }) = _ShuttleRoute;

  factory ShuttleRoute.fromJson(Map<String, dynamic> json) => _$ShuttleRouteFromJson(json);
}

@freezed
class ShuttleSchedule with _$ShuttleSchedule {
  const factory ShuttleSchedule({
    @JsonKey(name: 'depart_at') required DateTime departAt,
    @JsonKey(name: 'arrive_at') DateTime? arriveAt,
  }) = _ShuttleSchedule;

  factory ShuttleSchedule.fromJson(Map<String, dynamic> json) => _$ShuttleScheduleFromJson(json);
}

@freezed
class ShuttleCapacity with _$ShuttleCapacity {
  const factory ShuttleCapacity({
    required int total,
    required int taken,
    required int remaining,
  }) = _ShuttleCapacity;

  factory ShuttleCapacity.fromJson(Map<String, dynamic> json) => _$ShuttleCapacityFromJson(json);
}

enum ShuttleStatus {
  @JsonValue('open')
  open,
  @JsonValue('in_progress')
  inProgress,
  @JsonValue('done')
  done,
  @JsonValue('canceled')
  canceled,
}

enum ShuttlePurpose {
  evacuation,
  medical_transport,
  supply_transport,
  volunteer_shuttle,
  other,
}

enum ShuttleCostType {
  @JsonValue('free')
  free,
  @JsonValue('share_gas')
  shareGas,
}

@freezed
class Shuttle with _$Shuttle {
  const factory Shuttle({
    required String id,
    required LocalizedText title,
    LocalizedText? description,
    @JsonKey(name: 'vehicle_info') required ShuttleVehicleInfo vehicleInfo,
    @Default(<ShuttlePurpose>[]) List<ShuttlePurpose> purposes,
    required ShuttleRoute route,
    required ShuttleSchedule schedule,
    required ShuttleCapacity capacity,
    @JsonKey(name: 'cost_type') @Default(ShuttleCostType.free) ShuttleCostType costType,
    @Default(ShuttleStatus.open) ShuttleStatus status,
    @Default(false) bool priority,
    @JsonKey(name: 'contact_name') String? contactName,
    @JsonKey(name: 'contact_phone_masked') String? contactPhoneMasked,
    @JsonKey(name: 'creator_id') required String creatorId,
    @JsonKey(name: 'chat_room_id') String? chatRoomId,
    @JsonKey(name: 'participants_snapshot') @Default(<String>[]) List<String> participantsSnapshot,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _Shuttle;

  factory Shuttle.fromJson(Map<String, dynamic> json) => _$ShuttleFromJson(json);
}

@freezed
class ShuttleParticipant with _$ShuttleParticipant {
  const factory ShuttleParticipant({
    @JsonKey(name: 'shuttle_id') required String shuttleId,
    @JsonKey(name: 'user_id') required String userId,
    @Default('passenger') String role,
    @JsonKey(name: 'is_visible') @Default(true) bool isVisible,
    @JsonKey(name: 'joined_at') DateTime? joinedAt,
  }) = _ShuttleParticipant;

  factory ShuttleParticipant.fromJson(Map<String, dynamic> json) => _$ShuttleParticipantFromJson(json);
}

@freezed
class ShuttleMessage with _$ShuttleMessage {
  const factory ShuttleMessage({
    required String id,
    @JsonKey(name: 'shuttle_id') required String shuttleId,
    @JsonKey(name: 'author_id') required String authorId,
    required String content,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'expires_at') DateTime? expiresAt,
  }) = _ShuttleMessage;

  factory ShuttleMessage.fromJson(Map<String, dynamic> json) => _$ShuttleMessageFromJson(json);
}

enum ResourceStatus {
  @JsonValue('active')
  active,
  @JsonValue('shortage')
  shortage,
  @JsonValue('closed')
  closed,
}

enum ResourceCategory { food, water, medical, shelter, power, other }

@freezed
class ResourcePoint with _$ResourcePoint {
  const factory ResourcePoint({
    required String id,
    required LocalizedText title,
    LocalizedText? description,
    @Default(<ResourceCategory>[]) List<ResourceCategory> categories,
    @JsonKey(name: 'status') @Default(ResourceStatus.active) ResourceStatus status,
    @JsonKey(name: 'expiry') DateTime? expiry,
    required GeoLocation location,
    String? address,
    @JsonKey(name: 'contact_masked_phone') String? contactMaskedPhone,
    @JsonKey(name: 'created_by') required String createdBy,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _ResourcePoint;

  factory ResourcePoint.fromJson(Map<String, dynamic> json) => _$ResourcePointFromJson(json);
}

@freezed
class ChatRoom with _$ChatRoom {
  const factory ChatRoom({
    required String id,
    @JsonKey(name: 'entity_type') required String entityType,
    @JsonKey(name: 'entity_id') required String entityId,
    @JsonKey(name: 'expires_at') DateTime? expiresAt,
    @JsonKey(name: 'created_by') String? createdBy,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _ChatRoom;

  factory ChatRoom.fromJson(Map<String, dynamic> json) => _$ChatRoomFromJson(json);
}

@freezed
class NotificationEvent with _$NotificationEvent {
  const factory NotificationEvent({
    required int id,
    @JsonKey(name: 'entity_type') required String entityType,
    @JsonKey(name: 'entity_id') required String entityId,
    required NotificationPayload payload,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _NotificationEvent;

  factory NotificationEvent.fromJson(Map<String, dynamic> json) => _$NotificationEventFromJson(json);
}

@freezed
class NotificationPayload with _$NotificationPayload {
  const factory NotificationPayload({
    @JsonKey(name: 'changed_fields') @Default(<String>[]) List<String> changedFields,
    required String status,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _NotificationPayload;

  factory NotificationPayload.fromJson(Map<String, dynamic> json) => _$NotificationPayloadFromJson(json);
}

File: disaster_relief_app/docs/drip_models.ts
// DRIP shared models (TypeScript) aligned with specs_arch.md & specs_feature.md
export type LocaleKey = 'zh-TW' | 'en-US';
export type MissionStatus = 'open' | 'in_progress' | 'done' | 'canceled';
export type ShuttleStatus = 'open' | 'in_progress' | 'done' | 'canceled';
export type ShuttleCostType = 'free' | 'share_gas';
export type ResourceStatus = 'active' | 'shortage' | 'closed';

export type MissionType = 'rescue' | 'medical' | 'transport' | 'clearing' | 'other';
export type ShuttlePurpose = 'evacuation' | 'medical_transport' | 'supply_transport' | 'volunteer_shuttle' | 'other';
export type ResourceCategory = 'food' | 'water' | 'medical' | 'shelter' | 'power' | 'other';

export interface LocalizedText {
  'zh-TW': string;
  'en-US': string;
}

export interface GeoLocation {
  lat: number;
  lng: number;
  address: string;
  geohash?: string | null;
}

export interface MissionRequirements {
  materials: string[];
  manpower_needed: number;
  manpower_current: number;
}

export interface Mission {
  id: string;
  title: LocalizedText;
  description?: LocalizedText | null;
  types: MissionType[];
  status: MissionStatus;
  priority: boolean;
  requirements: MissionRequirements;
  location: GeoLocation;
  contact_name?: string | null;
  contact_phone_masked?: string | null;
  creator_id: string;
  chat_room_id?: string | null;
  participant_count: number;
  created_at: string;
  updated_at: string;
}

export interface MissionParticipant {
  mission_id: string;
  user_id: string;
  is_visible: boolean;
  joined_at: string;
}

export interface MissionMessage {
  id: string;
  mission_id: string;
  author_id: string;
  content: string;
  created_at: string;
  expires_at: string;
}

export interface ShuttleVehicleInfo {
  type: 'bus' | 'van' | 'car' | 'truck' | 'other';
  plate?: string;
}

export interface ShuttleRoute {
  origin: GeoLocation;
  destination: GeoLocation;
  stops?: GeoLocation[];
}

export interface ShuttleSchedule {
  depart_at: string;
  arrive_at?: string | null;
}

export interface ShuttleCapacity {
  total: number;
  taken: number;
  remaining: number;
}

export interface Shuttle {
  id: string;
  title: LocalizedText;
  description?: LocalizedText | null;
  vehicle_info: ShuttleVehicleInfo;
  purposes: ShuttlePurpose[];
  route: ShuttleRoute;
  schedule: ShuttleSchedule;
  capacity: ShuttleCapacity;
  cost_type: ShuttleCostType;
  status: ShuttleStatus;
  priority: boolean;
  contact_name?: string | null;
  contact_phone_masked?: string | null;
  creator_id: string;
  chat_room_id?: string | null;
  participants_snapshot: string[];
  created_at: string;
  updated_at: string;
}

export interface ShuttleParticipant {
  shuttle_id: string;
  user_id: string;
  role: 'driver' | 'staff' | 'passenger';
  is_visible: boolean;
  joined_at: string;
}

export interface ShuttleMessage {
  id: string;
  shuttle_id: string;
  author_id: string;
  content: string;
  created_at: string;
  expires_at: string;
}

export interface ResourcePoint {
  id: string;
  title: LocalizedText;
  description?: LocalizedText | null;
  categories: ResourceCategory[];
  status: ResourceStatus;
  expiry: string;
  location: GeoLocation;
  address?: string | null;
  contact_masked_phone?: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface ChatRoom {
  id: string;
  entity_type: 'mission' | 'shuttle';
  entity_id: string;
  expires_at: string;
  created_by?: string | null;
  created_at: string;
}

export interface NotificationEvent {
  id: number;
  entity_type: 'mission' | 'shuttle';
  entity_id: string;
  payload: {
    changed_fields: string[];
    status: MissionStatus | ShuttleStatus;
    updated_at: string;
  };
  created_at: string;
}
